tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,284 @@
+        tolerance for the procedure. If the supremum norm of the gradient
+        vector is below this number, the algorithm is stopped.
+      name: (Optional) Python str. The name prefixed to the ops created
+        by this function. If not supplied, the default name 'minimize'
+        is used.
+
+    Returns:
+      optimizer_results: A RotosolveOptimizerResults object contains the
+       result of the optimization process.
+    """
+
+    with tf.name_scope(name or 'minimize'):
+        initial_position = tf.convert_to_tensor(initial_position,
+                                                name='initial_position',
+                                                dtype='float32')
+        dtype = initial_position.dtype.base_dtype
+        tolerance = tf.convert_to_tensor(tolerance,
+                                         dtype=dtype,
+                                         name='grad_tolerance')
+        max_iterations = tf.convert_to_tensor(max_iterations,
+                                              name='max_iterations')
+
+        def _rotosolve_one_parameter_once(state):
+            """
+            Rotosolve a single parameter.
+            """
+            delta_shift = tf.reshape(
+                tf.cast(tf.sparse.to_dense(
+                    tf.sparse.SparseTensor(
+                        [[state.solve_param_i, 0]], [math.pi / 2],
+                        [prefer_static_shape(state.position)[0], 1])),
+                        dtype=dtype), prefer_static_shape(state.position))
+
+            # Evaluate three different point for curve fitting
+            v_l, v_n, v_r = expectation_value_function(
+                state.position - delta_shift), \
+                state.objective_value, \
+                expectation_value_function(state.position + delta_shift)
+
+            # Use the analytical solution to find the optimized position
+            delta_update = -math.pi / 2 - \
+                tf.math.atan2(2 * v_n - v_l - v_r, v_r - v_l)
+
+            delta_update_tensor = tf.reshape(
+                tf.cast(tf.sparse.to_dense(
+                    tf.sparse.SparseTensor(
+                        [[state.solve_param_i, 0]], [delta_update],
+                        [prefer_static_shape(state.position)[0], 1])),
+                        dtype=dtype), prefer_static_shape(state.position))
+
+            state.solve_param_i.assign_add(1)
+            state.position.assign(
+                tf.math.floormod(state.position + delta_update_tensor,
+                                 math.pi * 2))
+
+            state.last_objective_value.assign(state.objective_value)
+            state.objective_value.assign(
+                expectation_value_function(state.position))
+
+            return [state]
+
+        def _rotosolve_all_parameters_once(state):
+            """
+            Loop iterate over all parameters and update each one of them
+            """

```

Comment:
ditto.

Reply:
Fixed in 64f76c59d428d1501648637c7f1cc3af399cc7ed